### PR TITLE
Interactive forms: render choice widget annotations

### PR DESCRIPTION
**Pull request by timvandermeij**
This patch is another part of #7613.
- The `readOnly` property has been moved to the `WidgetAnnotation` class as the specification says that all types of widget annotations may use this flag, and indeed we need it for choice widget annotations as well.
- The `fieldValue` property is now only fetched in the `WidgetAnnotation` class and verification is done in the inheriting classes. This is required because for text widget annotations it's a string, whereas for choice widget annotations it may be an array as well.

### CODE REVIEW

**Comment by timvandermeij** [comment]
The bots received an update to Firefox 51, which causes a lot of reference test changes. I do wonder, however, why the Windows bot did not do anything...

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: To prevent strange, and hard to find, bugs later on you probably want to use `getArray` here (or maybe `var option = params.xref.fetchIfRef(options[i]);` below).

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
I'll try to do a proper review later in the week!
Taking a quick look at the PR I've noticed two things so far, please see the inline comment and the error below.

When opening `annotation-choice-widget.pdf` locally against `master` with forms _disabled_, I get the following error:

```
An error occurred while rendering the page.
PDF.js v1.5.488 (build: 6c263c1)
Message: str.charCodeAt is not a function
Stack: stringToPDFString@resource://pdf.js/build/pdf.worker.js:2035:42
WidgetAnnotation@resource://pdf.js/build/pdf.worker.js:40776:23
AnnotationFactory_create@resource://pdf.js/build/pdf.worker.js:40266:16
get annotations@resource://pdf.js/build/pdf.worker.js:41471:26
LocalPdfManager_ensure/<@resource://pdf.js/build/pdf.worker.js:41862:15
LocalPdfManager_ensure@resource://pdf.js/build/pdf.worker.js:41860:14
Page_getOperatorList@resource://pdf.js/build/pdf.worker.js:41389:32
wphSetupRenderPage/<@resource://pdf.js/build/pdf.worker.js:42660:9
Async*wphSetupRenderPage@resource://pdf.js/build/pdf.worker.js:42653:7
messageHandlerComObjOnMessage@resource://pdf.js/build/pdf.worker.js:2607:9
EventListener.handleEvent*MessageHandler@resource://pdf.js/build/pdf.worker.js:2613:3
wphCreateDocumentHandler@resource://pdf.js/build/pdf.worker.js:42301:19
wphSetupDoc@resource://pdf.js/build/pdf.worker.js:42288:14
messageHandlerComObjOnMessage/<@resource://pdf.js/build/pdf.worker.js:2584:18
Async*messageHandlerComObjOnMessage@resource://pdf.js/build/pdf.worker.js:2583:9
EventListener.handleEvent*MessageHandler@resource://pdf.js/build/pdf.worker.js:2613:3
initializeWorker@resource://pdf.js/build/pdf.worker.js:42812:17
pdfjsWrapper/<@resource://pdf.js/build/pdf.worker.js:42820:3
pdfjsWrapper/<@resource://pdf.js/build/pdf.worker.js:41975:5
pdfjsWrapper@resource://pdf.js/build/pdf.worker.js:41973:2
@resource://pdf.js/build/pdf.worker.js:40:13
@resource://pdf.js/build/pdf.worker.js:25:1
@resource://pdf.js/build/pdf.worker.js:18:2
```

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Good find. I'll fix that when the comments from the full review are also present.

**Comment by timvandermeij** [comment]
The error is in fact a bug in the current `master` branch. As far as I can tell, it has been there since the moment the original widget code has been added. The problem is that https://github.com/mozilla/pdf.js/pull/7671/files#diff-18ca06e2bd5be4c9132b51f78388b8f0L622 expects the `V` entry to be a string, but as you also mentioned before, in the case of choice widgets it's an array, which leads to problems as `stringToPDFString` cannot handle that. This patch fixes that by moving the `stringToPDFString` conversion into `TextWidgetAnnotation`.

**Comment by benweet** [comment]
I've just tested this, it works beautifully. You can merge this before #7664 if you want, I can rebase and merge after that.

**Comment by timvandermeij** [comment]
@Snuffleupagus Do you have more review comments or can I start fixing this up with your current comments so we can land this?

**Comment by Snuffleupagus** [comment]
> @Snuffleupagus Do you have more review comments or can I start fixing this up with your current comments so we can land this?

Really sorry for the delay here!
I had intended to test/review this properly during the weekend, but after seeing [this `dev.planning` thread late on Friday](https://groups.google.com/forum/#!topic/mozilla.dev.planning/j834iDIG3yY) I had a difficult time mustering up the energy required to do a proper review.

However, I'll try to review this either tonight or tomorrow; again sorry for the delay here!

File: test/unit/annotation_layer_spec.js
Comment (Snuffleupagus) [inline test/unit/annotation_layer_spec.js]: Here, and in the other tests, you may also want to explicitly check that you actually get the expected annotation type, e.g. like this:
```javascript
expect(data.annotationType).toEqual(AnnotationType.WIDGET);
```

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: Similar to my other comment about references inside of arrays, if `this.data.fieldValue` already is an array you will have problems in the `display` layer if an array element is a `Ref`.

Please refer to the comment on the `data.fieldValue = Util.getInheritableProperty(dict, 'V');` line for a suggested solution.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: Since the `V` entry can be an array, you probably need to modify `Util.getInheritableProperty` to handle that. I can see two possible solutions:
 - Modify `Util.getInheritableProperty` to use always use `Dict_getArray` instead of `Dict_get`. Even though it only looks like this utility function is used from `src/core/annotation.js`, I'm not sure that this is such a great solution!
 - Add a parameter to `Util.getInheritableProperty` that controls whether `Dict_getArray` is used, e.g. something like this:
```diff
diff --git a/src/core/annotation.js b/src/core/annotation.js
index 13aa204..d7ef2ca 100644
--- a/src/core/annotation.js
+++ b/src/core/annotation.js
@@ -619,8 +619,8 @@ var WidgetAnnotation = (function WidgetAnnotationClosure() {
     var data = this.data;
 
     data.annotationType = AnnotationType.WIDGET;
-    data.fieldValue = stringToPDFString(
-      Util.getInheritableProperty(dict, 'V') || '');
+    data.fieldValue = Util.getInheritableProperty(dict, 'V',
+                                                  /* getArray = */ true);
     data.alternativeText = stringToPDFString(dict.get('TU') || '');
     data.defaultAppearance = Util.getInheritableProperty(dict, 'DA') || '';
     var fieldType = Util.getInheritableProperty(dict, 'FT');
diff --git a/src/shared/util.js b/src/shared/util.js
index 48876da..614e7a0 100644
--- a/src/shared/util.js
+++ b/src/shared/util.js
@@ -874,9 +874,10 @@ var Util = (function UtilClosure() {
   };
 
   Util.getInheritableProperty = function Util_getInheritableProperty(dict,
-                                                                     name) {
+                                                                     name,
+                                                                     getArray) {
     while (dict && !dict.has(name)) {
-      dict = dict.get('Parent');
+      dict = getArray ? dict.getArray('Parent') : dict.get('Parent');
     }
     if (!dict) {
       return null;

```

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js]: When we only supported `TextWidgetAnnotation`s, I suppose it didn't matter too much that we set the `isRenderable` parameter inside of `WidgetAnnotationElement`, however now I don't think that's correct anymore!
*Note:* For `TextWidgetAnnotation` there's a meaningful fallback in the case where no `Appearance` stream existed, and was thus not rendered in the `core` layer, but for the `ChoiceWidgetAnnotationElement` we're not rendering anything in the `display` layer by default.

I believe that we should to modify the current code like below, which will help not only here but when implementing other widget's later.
```diff
diff --git a/src/display/annotation_layer.js b/src/display/annotation_layer.js
index 29ed79a..76a8b73 100644
--- a/src/display/annotation_layer.js
+++ b/src/display/annotation_layer.js
@@ -400,9 +400,7 @@ var TextAnnotationElement = (function TextAnnotationElementClosure() {
  * @alias WidgetAnnotationElement
  */
 var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
-  function WidgetAnnotationElement(parameters) {
-    var isRenderable = parameters.renderInteractiveForms ||
-      (!parameters.data.hasAppearance && !!parameters.data.fieldValue);
+  function WidgetAnnotationElement(parameters, isRenderable) {
     AnnotationElement.call(this, parameters, isRenderable);
   }
 
@@ -432,7 +430,9 @@ var TextWidgetAnnotationElement = (
   var TEXT_ALIGNMENT = ['left', 'center', 'right'];
 
   function TextWidgetAnnotationElement(parameters) {
-    WidgetAnnotationElement.call(this, parameters);
+    var isRenderable = parameters.renderInteractiveForms ||
+      (!parameters.data.hasAppearance && !!parameters.data.fieldValue);
+    WidgetAnnotationElement.call(this, parameters, isRenderable);
   }
 
   Util.inherit(TextWidgetAnnotationElement, WidgetAnnotationElement, {
```

Then this line becomes
```javascript
WidgetAnnotationElement.call(this, parameters, parameters.renderInteractiveForms);
```
and the `if (this.renderInteractiveForms) {` condition is no longer necessary inside of `render`.

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js]: This `if` can be removed with the above comment addressed.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Again, sorry for the delay here! The PR itself looks good, however it exposes a couple of _existing_ issues in the code; please see the inline comments.

r=me, with comments addressed and passing tests.

**Review by Snuffleupagus — APPROVED** [review]

**Pull request merged**
